File: backend/internal/domain/house.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

type HouseTokenStatus string

const (
	HouseTokenActive    HouseTokenStatus = "active"
	HouseTokenRevoked   HouseTokenStatus = "revoked"
	HouseTokenSuspended HouseTokenStatus = "suspended"
)

type House struct {
	ID             uuid.UUID        `json:"id"`
	BlockNumber    string           `json:"block_number"`
	Address        *string          `json:"address,omitempty"`
	HeadResidentID *uuid.UUID       `json:"head_resident_id,omitempty"`
	HeadResident   *Resident        `json:"head_resident,omitempty"`
	UserID         *uuid.UUID       `json:"user_id,omitempty"`
	AccessToken    string           `json:"access_token"`
	TokenStatus    HouseTokenStatus `json:"token_status"`
	PinCode        string           `json:"pin_code"`
	TokenVersion   int              `json:"token_version"`
	CreatedAt      time.Time        `json:"created_at"`
	UpdatedAt      time.Time        `json:"updated_at"`
}

type HouseAccessClaimResponse struct {
	Token        string    `json:"token"`
	House        House     `json:"house"`
	TenantSlug   string    `json:"tenant_slug"`
	TenantName   string    `json:"tenant_name"`
	HeadResident *Resident `json:"head_resident,omitempty"`
	User         *User     `json:"user,omitempty"`
}

type HouseRepository interface {
	Create(ctx context.Context, tenantID uuid.UUID, house *House) error
	GetByID(ctx context.Context, tenantID, id uuid.UUID) (*House, error)
	GetByToken(ctx context.Context, tenantID uuid.UUID, token string) (*House, error)
	GetByUserID(ctx context.Context, tenantID, userID uuid.UUID) (*House, error)
	List(ctx context.Context, tenantID uuid.UUID, limit, offset int) ([]House, int, error)
	Update(ctx context.Context, tenantID uuid.UUID, house *House) error
	RevokeAndRegenerateToken(ctx context.Context, tenantID, id uuid.UUID, newToken, newPin string) error
	ResetPin(ctx context.Context, tenantID, id uuid.UUID, newPin string) error
	Delete(ctx context.Context, tenantID, id uuid.UUID) error
}

type HouseUsecase interface {
	ClaimAccessToken(ctx context.Context, tenantSlug, token, pin string) (*HouseAccessClaimResponse, error)
	GetMyHouse(ctx context.Context, tenantID, userID uuid.UUID) (*House, *Resident, error)
	ListHouses(ctx context.Context, tenantID uuid.UUID, limit, offset int) ([]House, int, error)
	CreateHouse(ctx context.Context, tenantID uuid.UUID, house *House) (*House, error)
	UpdateHouse(ctx context.Context, tenantID uuid.UUID, house *House) (*House, error)
	DeleteHouse(ctx context.Context, tenantID, houseID uuid.UUID) error
	RegenerateToken(ctx context.Context, tenantID, houseID uuid.UUID) (*House, error)
	ResetPin(ctx context.Context, tenantID, houseID uuid.UUID) (*House, error)
	VerifyPin(ctx context.Context, tenantID, houseID uuid.UUID, inputPin string) (bool, error)
}

File: frontend/src/services/house.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { api } from './api';
import { Resident } from '../types/resident';

export interface House {
  id: string;
  block_number: string;
  address?: string;
  head_resident_id?: string;
  head_resident?: Resident;
  user_id?: string;
  access_token: string;
  token_status: 'active' | 'revoked' | 'suspended';
  pin_code?: string;
  token_version?: number;
  created_at: string;
  updated_at: string;
}

export interface HouseListResponse {
  data: House[];
  total: number;
}

export interface CreateHousePayload {
  block_number: string;
  address?: string;
  head_resident_id?: string;
}

export interface ClaimHouseTokenResponse {
  token: string;
  house: House;
  tenant_slug: string;
  tenant_name: string;
  head_resident?: Resident;
  user?: {
    id: string;
    name: string;
    email: string;
    role?: string;
  };
}

export function useHouses(params?: { limit?: number; offset?: number }) {
  return useQuery({
    queryKey: ['admin-houses', params],
    queryFn: async () => {
      const res = await api.get<HouseListResponse>('/admin/houses', {
        params: { limit: params?.limit || 50, offset: params?.offset || 0 },
      });
      return res.data;
    },
  });
}

export function useCreateHouse() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (payload: CreateHousePayload) => {
      const res = await api.post<House>('/admin/houses', payload);
      return res.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['admin-houses'] });
    },
  });
}

export function useUpdateHouse() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, ...payload }: CreateHousePayload & { id: string }) => {
      const res = await api.put<House>(`/admin/houses/${id}`, payload);
      return res.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['admin-houses'] });
    },
  });
}

export function useDeleteHouse() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (houseId: string) => {
      const res = await api.delete<{ message: string }>(`/admin/houses/${houseId}`);
      return res.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['admin-houses'] });
    },
  });
}

export function useRegenerateHouseToken() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (houseId: string) => {
      const res = await api.post<House>(`/admin/houses/${houseId}/regenerate-token`);
      return res.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['admin-houses'] });
    },
  });
}

export function useResetHousePin() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (houseId: string) => {
      const res = await api.post<House>(`/admin/houses/${houseId}/reset-pin`);
      return res.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['admin-houses'] });
    },
  });
}

export async function claimHouseToken(slug: string, token: string): Promise<ClaimHouseTokenResponse> {
  const res = await api.get<ClaimHouseTokenResponse>('/house-access/claim', {
    params: { slug, token },
  });
  return res.data;
}
